MantidPlot/src/ProjectRecoveryGUIs/ProjectRecoveryModel.cpp

```diff
@@ -0,0 +1,219 @@
+  createThreadAndManage(checkpoint);
+
+  selected.replace(selected.find("T"), 1, " ");
+  if (m_failedRun) {
+    updateCheckpointTried(selected);
+  }
+  m_recoveryRunning = false;
+  // Close View
+  m_presenter->closeView();
+}
+void ProjectRecoveryModel::openSelectedInEditor(std::string &selected) {
+  m_recoveryRunning = true;
+  // Clear the ADS
+  Mantid::API::AnalysisDataService::Instance().clear();
+
+  // Open editor for this checkpoint
+  auto stringSpacePos = selected.find(" ");
+  if (stringSpacePos != std::string::npos)
+    selected.replace(stringSpacePos, 1, "T");
+  auto beforeCheckpoint = m_projRec->getRecoveryFolderLoadPR();
+  Poco::Path checkpoint(beforeCheckpoint);
+  checkpoint.append(selected);
+  Poco::Path output(Mantid::Kernel::ConfigService::Instance().getAppDataDir());
+  output.append("ordered_recovery.py");
+
+  m_projRec->openInEditor(checkpoint, output);
+  // Restart project recovery as we stay synchronous
+  m_projRec->clearAllCheckpoints(beforeCheckpoint);
+  m_projRec->startProjectSaving();
+
+  selected.replace(selected.find("T"), 1, " ");
+  if (m_failedRun) {
+    updateCheckpointTried(selected);
+  }
+  m_recoveryRunning = false;
+  m_failedRun = false;
+  // Close View
+  m_presenter->closeView();
+}
+
+void ProjectRecoveryModel::fillRow(Poco::Path path,
+                                   std::string checkpointName) {
+  std::string lengthOfFile = findNumberOfWorkspacesInDirectory(path);
+  std::string checked = "No";
+  std::vector<std::string> nextVector = {checkpointName, lengthOfFile, checked};
+  m_rows.emplace_back(nextVector);
+}
+
+void ProjectRecoveryModel::fillFirstRow() {
+  auto paths = m_projRec->getListOfFoldersInDirectoryPR(
+      m_projRec->getRecoveryFolderLoadPR());
+
+  std::sort(paths.begin(), paths.end(), [](auto &a, auto &b) {
+    Poco::File a1(a);
+    Poco::File b1(b); // Last modified is first!
+    return a1.getLastModified() > b1.getLastModified();
+  });
+
+  // Grab the last path as that is the one that should be loaded
+  auto path = paths[0];
+  std::string checkpointName = path.directory(path.depth() - 1);
+  checkpointName.replace(checkpointName.find("T"), 1, " ");
+  fillRow(path, checkpointName);
+}
+
+void ProjectRecoveryModel::fillRows() {
+  auto paths = m_projRec->getListOfFoldersInDirectoryPR(
+      m_projRec->getRecoveryFolderLoadPR());
+
+  std::sort(paths.begin(), paths.end(), [](auto &a, auto &b) {
+    Poco::File a1(a);
+    Poco::File b1(b); // Last modified is first!
+    return a1.getLastModified() > b1.getLastModified();
+  });
+
+  // Sort the rows first string of the vector lists
+  for (auto c : paths) {
+    std::string checkpointName = c.directory(c.depth() - 1);
+    checkpointName.replace(checkpointName.find("T"), 1, " ");
+    // Check if there is a first row already and skip first one
+    if (m_rows[0][0] == checkpointName)
+      continue;
+
+    fillRow(c, checkpointName);
+  }
+
+  for (auto i = paths.size(); i < 5; ++i) {
```

Comment:
What's the magic number 5? Could it have some nicer name as `constexpr size_t ...`?

Reply:
It's actually meant to be a somewhat dynamic number I mean to change after initial implementation good spot! :+1: